### ISSUE TITLE
street name not showing up

### ISSUE THREAD

**Issue by iriman** [gravitystorm/openstreetmap-carto#4532]
### Links and screenshots illustrating the problem

https://www.openstreetmap.org/way/141004128

![image](https://user-images.githubusercontent.com/8467903/163165777-0ea7b704-1792-49ef-8e99-b07e1287c021.png)

**Comment by sommerluk** [comment]
What a complicate issue!

- The way https://www.openstreetmap.org/way/141004128 which you are saying is not showing up, has been edited in the meantime, and the U+200C ZERO WIDTH NON-JOINER has been replaced with U+0020 SPACE. Therefore, I have created a local, modified version with the old name like in Version 26 of this way.
- I cannot reproduce this behaviour locally neither with the more-or-less up-to-date 39873d2f2c212b63ea8d849b35a4f190f5ea6aa2 nor with the older 4b58c59093f3973566ad8f0ee679560e8c6c8d75 (which should be more or less what is running on openstreetmap.org currently). Instead, for me, it shows up, but with many tofu characters (these strange rectangles): ![grafik](https://user-images.githubusercontent.com/6830724/187049410-92fc5972-20fe-4b93-aaa3-d8fddd12f568.png) So the results are not identical on different computers.
- Note that Mapnik [describes](http://mapnik.org/mapnik-reference/#3.0.22/font-directory) the font directory parameter as follows: _Path to a directory which holds fonts which should be registered when the Map is loaded (in addition to any fonts that may be automatically registered)._ That means that even now that we are vendoring fonts, we cannot be sure to have the same result on any setup, because we do not control the installed system fonts, and these installed system fonts continue to be used.
- You mention that the way https://www.openstreetmap.org/way/319229037 which also contains U+200C ZERO WIDTH NON-JOINER does show up correctly on openstreetmap.org. Indeed, it does. But also this behaviour I cannot reproduce locally. On my local machine, I get the same tofu as for https://www.openstreetmap.org/way/141004128 in version 26: 
![grafik](https://user-images.githubusercontent.com/6830724/187049803-89010a1a-3dc6-4dc0-831b-4af35ad77ebd.png)
- Indeed, removing the U+200C ZERO WIDTH NON-JOINER character makes the issue disappear.
- Modifying the font list in `fonts.mss` so that Noto Sans Arabic Regular is the first one, also makes the issue disappear.
- The U+200C ZERO WIDTH NON-JOINER character has special semantics. Its glyph should never show up. The rendering engine must implement this special behaviour. If it does not implement this special behaviour, [as a fallback solution, it should simply ignore this character.](http://unicode.org/faq/unsup_char.html#3) By the way, Noto Sans Arabic UI Regular includes such a glyph, showing the [ISO keyboard symbol for ZWNJ](https://en.wikipedia.org/wiki/File:ISOIEC-9995-7-081--IEC-60417-6077-1--Symbol-for-ZWNJ.svg) Noto Sans Balinese Regular has rather an empty glyph. And Noto Sans Regular I could not open in FontForge to verify.

In summary, I think this is a bug in the font rendering stack of Mapnik, either in Mapnik or in the underlying libraries. (Note that Mapnik renders badly the Noto Sans Adlam joined script on my local machine, while other applications render it correctly, so I suspect Mapnik does not simply use the operating system's HarfBuzz library, but does something special.)

In this repository, I think we cannot do anything to solve this issue.

**Comment by pnorman** [comment]
> I suspect Mapnik does not simply use the operating system's HarfBuzz library, but does something special

It depends how you've compiled it.

**Comment by sommerluk** [comment]
Closing the issue here in openstreetmap-carto. It’s an upstream issue. There is yet an issue in Mapnik covering this: https://github.com/mapnik/mapnik/issues/3931
